polaris-core/src/main/java/org/apache/polaris/core/storage/s3compatible/S3CompatibleCredentialsStorageIntegration.java

```diff
@@ -0,0 +1,164 @@
+    EnumMap<PolarisCredentialProperty, String> propertiesMap =
+        new EnumMap<>(PolarisCredentialProperty.class);
+    propertiesMap.put(PolarisCredentialProperty.AWS_ENDPOINT, storageConfig.getS3Endpoint());
+    propertiesMap.put(
+        PolarisCredentialProperty.AWS_PATH_STYLE_ACCESS,
+        storageConfig.getS3PathStyleAccess().toString());
+
+    switch (storageConfig.getCredsVendingStrategy()) {
+      case KEYS_SAME_AS_CATALOG:
+        if (storageConfig.getCredsCatalogAndClientStrategy()
+            == CredsCatalogAndClientStrategyEnum.ENV_VAR_NAME) {
+          String cai = System.getenv(storageConfig.getS3CredentialsCatalogAccessKeyId());
+          String cas = System.getenv(storageConfig.getS3CredentialsCatalogSecretAccessKey());
+          if (cai == null || cas == null) {
+            throw new IllegalArgumentException(
+                "Expect valid environment variable for catalog keys");
+          } else {
+            propertiesMap.put(PolarisCredentialProperty.AWS_KEY_ID, cai);
+            propertiesMap.put(PolarisCredentialProperty.AWS_SECRET_KEY, cas);
+          }
+        } else {
+          propertiesMap.put(
+              PolarisCredentialProperty.AWS_KEY_ID,
+              storageConfig.getS3CredentialsCatalogAccessKeyId());
+          propertiesMap.put(
+              PolarisCredentialProperty.AWS_SECRET_KEY,
+              storageConfig.getS3CredentialsCatalogSecretAccessKey());
+        }
+        break;
+
+      case KEYS_DEDICATED_TO_CLIENT:
+        if (storageConfig.getCredsCatalogAndClientStrategy()
+            == CredsCatalogAndClientStrategyEnum.ENV_VAR_NAME) {
+          String cli = System.getenv(storageConfig.getS3CredentialsClientAccessKeyId());
+          String cls = System.getenv(storageConfig.getS3CredentialsClientSecretAccessKey());
+          if (cli == null || cls == null) {
+            throw new IllegalArgumentException("Expect valid environment variable for client keys");
+          } else {
+            propertiesMap.put(PolarisCredentialProperty.AWS_KEY_ID, cli);
+            propertiesMap.put(PolarisCredentialProperty.AWS_SECRET_KEY, cls);
+          }
+        } else {
+          propertiesMap.put(
+              PolarisCredentialProperty.AWS_KEY_ID,
+              storageConfig.getS3CredentialsClientAccessKeyId());
+          propertiesMap.put(
+              PolarisCredentialProperty.AWS_SECRET_KEY,
+              storageConfig.getS3CredentialsClientSecretAccessKey());
+        }
+        break;
+
+      case TOKEN_WITH_ASSUME_ROLE:
+        if (this.stsClient == null) {
+          createStsClient(storageConfig);
+        }
+        LOGGER.debug("S3Compatible - assumeRole !");
+        AssumeRoleResponse response =
+            stsClient.assumeRole(
+                AssumeRoleRequest.builder()
+                    .roleSessionName("PolarisCredentialsSTS")
+                    // @TODO customize duration
+                    // .durationSeconds(1800)
```

Comment:
See:
https://github.com/apache/polaris/pull/408

Let's just make sure to do this in whichever PR merges second :)